plugins/extractors/metabase/extractor_test.go

```diff
@@ -0,0 +1,341 @@
+		ID string `json:"id"`
+	}
+	var data response
+	err = unmarshalResponse(res, &data)
+	if err != nil {
+		return
+	}
+	session_id = data.ID
+	err = getSessionID()
+	return
+}
+
+func getSessionID() (err error) {
+	values := map[string]interface{}{
+		"username": email,
+		"password": pass,
+	}
+	jsonValue, err := json.Marshal(values)
+	if err != nil {
+		return
+	}
+	res, err := newRequest("POST", url+"/api/session", bytes.NewBuffer(jsonValue))
+	if err != nil {
+		return
+	}
+	type response struct {
+		ID string `json:"id"`
+	}
+	var data response
+	err = unmarshalResponse(res, &data)
+	if err != nil {
+		return
+	}
+	session_id = data.ID
+	return
+}
+
+func addMockData(session_id string) (err error) {
+	err = addCollection()
+	if err != nil {
+		return
+	}
+	err = addDashboard()
+	if err != nil {
+		return
+	}
+	return
+}
+
+func addCollection() (err error) {
+	values := map[string]interface{}{
+		"name":        collection_name,
+		"color":       collection_color,
+		"description": collection_description,
+	}
+
+	jsonValue, err := json.Marshal(values)
+	if err != nil {
+		return
+	}
+	res, err := newRequest("POST", url+"/api/collection", bytes.NewBuffer(jsonValue))
+	if err != nil {
+		return
+	}
+	type response struct {
+		ID int `json:"id"`
+	}
+	var data response
+	err = unmarshalResponse(res, &data)
+	if err != nil {
+		return
+	}
+	collection_id = data.ID
+	return
+}
+
+func addDashboard() (err error) {
+	values := map[string]interface{}{
+		"name":          dashboard_name,
+		"description":   dashboard_description,
+		"collection_id": collection_id,
+	}
+	jsonValue, err := json.Marshal(values)
+	if err != nil {
+		return
+	}
+	res, err := newRequest("POST", url+"/api/dashboard", bytes.NewBuffer(jsonValue))
+	if err != nil {
+		return
+	}
+	type response struct {
+		ID int `json:"id"`
+	}
+	var data response
+	err = unmarshalResponse(res, &data)
+	if err != nil {
+		return
+	}
+	dashboard_id = data.ID
+	err = addCard(dashboard_id)
+	if err != nil {
+		return
+	}
+	return
+}
+
+func addCard(id int) (err error) {
+	values := map[string]interface{}{
+		"id": id,
+	}
+	jsonValue, err := json.Marshal(values)
+	if err != nil {
+		return
+	}
+	x := strconv.Itoa(id)
+	res, err := newRequest("POST", url+"/api/dashboard/"+x+"/cards", bytes.NewBuffer(jsonValue))
+	if err != nil {
+		return
+	}
+	type response struct {
+		ID int `json:"id"`
+	}
+	var data response
+	err = unmarshalResponse(res, &data)
+	if err != nil {
+		return
+	}
+	card_id = data.ID
+	return
+}
+
+func newRequest(method, url string, body io.Reader) (res *http.Response, err error) {
+	req, err := http.NewRequest(method, url, body)
+	if err != nil {
+		return
+	}
+	if body != nil {
+		req.Header.Set("Content-Type", "application/json")
+	}
+	if session_id != "" {
+		req.Header.Set("X-Metabase-Session", session_id)
+	}
+	res, err = client.Do(req)
+	if err != nil {
+		return
+	}
+	return
+}
+
+func unmarshalResponse(res *http.Response, data interface{}) (err error) {
```

Comment:
we can merge this function directly to `newRequest` (also we should rename it to `sendRequest` since we are immediately do the request).
e.g.
```
func sendRequest(method, url string, data interface{}) (err error) {}
```

Reply:
same applied to `unmarshalResponse` method in extractor

Reply:
Got it, but it will require 4 parameters right??

```func sendRequest(method, url string, body io.Reader, data interface{}) (err error) {}```

Reply:
we wouldnt need `body io.Reader` anymore.
basically the method will create a request for you and map the result to `data interface{}`.

`body io.Reader` is just the implementation of the method.